barneycarroll/patchinko
# Add typescript definitions

I've been working on this for a while! It has been a very fun though hard challenge. Want to thank @fivitti for his feedbacks ;-)

resolves #24

Tested and linted typescript definitions for TS ⩾ 3.5.

### Highlights

The root of typings are in types/patchinko, a requirement from dtslint.
The most important tests are located in `__tests__/explicit` and `__tests__/overlaoded`.

One big feature of those typings is to strictly enforce patchinko rules. As such, every do and don't from the readme has been carefuly tested and implemented.

Example with `__tests__/overlaoded`:

![Screenshot_20200218_081613](https://user-images.githubusercontent.com/3646758/74731548-f4f2a900-5226-11ea-9647-86c482638143.png)

`__tests__/explicit`:

![Screenshot_20200218_081804](https://user-images.githubusercontent.com/3646758/74731665-32efcd00-5227-11ea-9a7c-2770b712b479.png)

### Limitations

One limitation of those types however, is that the returned type of a `P(target, ...)` call will always be `typeof target`. Although it could be possible, in theory, to find the resulting type from operations such as deletions ore scoped replacements, I have found this incredibly difficult and wasn't certain the compiler could follow this level of complexity and computational burden. Therefore, I thought it would be an acceptable compromise to return the target type, given the vast majority of use-cases where one want such outcome.

One who needs to circumvent those limitations just has to cast patch target to unknown:

```typescript
P<unknown>(x, { foo: PS({ tada: "" }) });
```

### Testing

Run tests with `npm run test:ts` which uses dtslint in the background.
It does the following:

- Lint the sources according to DefinitelyTyped guidelines;
- Run any test file and check for $ExpectType and $ExpectError instructions, report compilation errors;
- Run one batch of test for each TypeScript supported version.
